sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakeFileInputStream.java

```diff
@@ -0,0 +1,91 @@
+import com.azure.storage.file.datalake.models.DataLakeRequestConditions;
+import com.azure.storage.file.datalake.models.DataLakeStorageException;
+import com.azure.storage.file.datalake.models.FileRange;
+import com.azure.storage.file.datalake.models.PathProperties;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+/**
+ * Provides an input stream to read a given Data Lake file resource.
+ */
+public final class DataLakeFileInputStream extends StorageInputStream {
+    /**
+     * Holds the reference to the file this stream is associated with.
+     */
+    private final DataLakeFileAsyncClient fileClient;
+
+    /**
+     * Holds the {@link DataLakeRequestConditions} object that represents the access conditions for the file.
+     */
+    private final DataLakeRequestConditions accessCondition;
+
+    /**
+     * Holds the {@link PathProperties} object that represents the file's properties.
+     */
+    private final PathProperties properties;
+
+    /**
+     * Initializes a new instance of the DataLakeFileInputStream class. Note that if {@code fileRangeOffset} is not {@code 0} or
+     * {@code fileRangeLength} is not {@code null}, there will be no content MD5 verification.
+     *
+     * @param fileClient A {@link DataLakeFileAsyncClient} object which represents the file that this stream is associated
+     * with.
+     * @param fileRangeOffset The offset of file data to begin stream.
+     * @param fileRangeLength How much data the stream should return after fileRangeOffset.
+     * @param chunkSize The size of the chunk to download.
+     * @param accessCondition An {@link DataLakeRequestConditions} object which represents the access conditions for the
+     * file.
+     * @throws DataLakeStorageException An exception representing any error which occurred during the operation.
+     */
+    DataLakeFileInputStream(final DataLakeFileAsyncClient fileClient, long fileRangeOffset, Long fileRangeLength,
+        int chunkSize, final DataLakeRequestConditions accessCondition, final PathProperties pathProperties)
+        throws DataLakeStorageException {
+        super(fileRangeOffset, fileRangeLength, chunkSize, pathProperties.getFileSize());
+
+        this.fileClient = fileClient;
+        this.accessCondition = accessCondition;
+        this.properties = pathProperties;
+    }
+
+    /**
+     * Dispatches a read operation of N bytes.
+     *
+     * @param readLength An <code>int</code> which represents the number of bytes to read.
+     */
+    @Override
+    protected synchronized ByteBuffer dispatchRead(final int readLength, final long offset) throws IOException {
+        try {
+            ByteBuffer currentBuffer = this.fileClient.readWithResponse(
+                new FileRange(offset, (long) readLength), null, this.accessCondition, false)
+                .flatMap(response -> FluxUtil.collectBytesInByteBufferStream(response.getValue()).map(ByteBuffer::wrap))
+                .block();
```

Comment:
The fileClient delegtes reads to blockblob client internally. 
Maybe we should delegate opening a stream to blockblob client as well and just make this class an adapter, i.e. keep reference to stream from block blob and just proxy calls there? If we were returning plain InputStream from API we could just return stream from blobclient, but since we expose some extra properties adapter would be needed.
I'd consider this - writing an adapter is easier than maintaining two versions of logic that works on bytebuffers and offsets.


This is what dotnet does https://github.com/Azure/azure-sdk-for-net/blob/3f38e290bfc1b1579baa4abf329a3861355796f1/sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileClient.cs#L3869-L3873

Or maybe we don't look at blobs and just return InputStream ?

Reply:
So most of the decisions were made with the intent to mirror the blob API shape. It's been expressed multiple times throughout this PR that the blob API for this is undesirable. If we make the decision to break from that design, we can return a class that holds a plain InputStream and the datalake properties separately, then use the block blob inputstream as the base implementation to wrap. Is this acceptable to people? @kasobol-msft @alzimmermsft  @gapra-msft @rickle-msft

Reply:
Example of what would be returned:
```java
public class DataLakeFileInputStream {
    public InputStream getInputStream(); // literally returns a BlockBlobInputStream instance
    public PathProperties getProperties();
}
```

Reply:
I like that idea. This should be called `DataLakeFileInputStreamResult` (or whatever matches <operation>Result pattern).